cmd/restic/cmd_rewrite.go

```diff
@@ -118,50 +133,97 @@ func rewriteSnapshot(ctx context.Context, repo *repository.Repository, sn *resti
 		return false, err
 	}
 
-	metadata, err := opts.Metadata.convert()
+	includeByNameFuncs, err := opts.IncludePatternOptions.CollectPatterns(Warnf)
+	if err != nil {
+		return false, err
+	}
 
+	metadata, err := opts.Metadata.convert()
 	if err != nil {
 		return false, err
 	}
 
 	var filter rewriteFilterFunc
+	var rewriteNode walker.NodeRewriteFunc
+	var rewriter *walker.TreeRewriter
+	var querySize walker.QueryRewrittenSizeFunc
+
+	if len(rejectByNameFuncs) > 0 || len(includeByNameFuncs) > 0 || opts.SnapshotSummary {
+		if len(rejectByNameFuncs) > 0 {
+			selectByName := func(nodepath string) bool {
+				for _, reject := range rejectByNameFuncs {
+					if reject(nodepath) {
+						return false
+					}
+				}
+				return true
+			}
 
-	if len(rejectByNameFuncs) > 0 {
-		selectByName := func(nodepath string) bool {
-			for _, reject := range rejectByNameFuncs {
-				if reject(nodepath) {
-					return false
+			rewriteNode = func(node *restic.Node, path string) *restic.Node {
+				if selectByName(path) {
+					return node
 				}
+				Verbosef("excluding %s\n", path)
+				return nil
+			}
+
+		} else if len(includeByNameFuncs) > 0 {
+			selectByName := func(nodepath string, node *restic.Node) bool {
+				for _, include := range includeByNameFuncs {
+					if node.Type == restic.NodeTypeDir {
+						// always include directories
+						return true
+					} else if node.Type == restic.NodeTypeFile {
+						ifun, childMayMatch := include(nodepath)
+						return ifun && childMayMatch
+					} else if node.Type == restic.NodeTypeSymlink {
+						includeLink, _ := followSymlink(nodepath, node, include)
```

Comment:
Symlinks should be treated like normal files in regard to include patterns. Everything else is unusable.